golem-worker-service-base/src/gateway_execution/gateway_session.rs

```diff
@@ -0,0 +1,225 @@
+// Copyright 2024 Golem Cloud
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use async_trait::async_trait;
+use std::collections::HashMap;
+use std::hash::Hash;
+use std::sync::Arc;
+use std::time::{Duration, Instant};
+use tokio::sync::Mutex;
+
+#[async_trait]
+pub trait GatewaySession {
+    async fn insert(
+        &self,
+        session_id: SessionId,
+        data_key: DataKey,
+        data_value: DataValue,
+    ) -> Result<(), String>;
+
+    async fn get(&self, session_id: &SessionId) -> Result<Option<SessionData>, String>;
+
+    async fn get_data_value(
+        &self,
+        session_id: SessionId,
+        data_key: DataKey,
+    ) -> Result<Option<DataValue>, String>;
+    async fn get_params(
+        &self,
+        session_id: SessionId,
+    ) -> Result<Option<HashMap<DataKey, DataValue>>, String>;
+}
+
+#[derive(Clone)]
+pub struct GatewaySessionStore(pub Arc<dyn GatewaySession + Send + Sync>);
+
+impl GatewaySessionStore {
+    pub fn in_memory(eviction_strategy: &EvictionStrategy) -> Self {
+        GatewaySessionStore(Arc::new(InMemoryGatewaySession::new(eviction_strategy)))
+    }
+}
+
+#[derive(Debug, Hash, PartialEq, Eq, Clone)]
+pub struct SessionId(pub String);
+
+#[derive(Debug, Hash, PartialEq, Eq, Clone)]
+pub struct DataKey(pub String);
+
+impl DataKey {
+    pub fn nonce() -> DataKey {
+        DataKey("nonce".to_string())
+    }
+
+    pub fn claims() -> DataKey {
+        DataKey("claims".to_string())
+    }
+
+    pub fn redirect_url() -> DataKey {
+        DataKey("redirect_url".to_string())
+    }
+}
+
+#[derive(Debug, Clone)]
+pub struct DataValue(pub serde_json::Value);
+
+impl DataValue {
+    pub fn as_string(&self) -> Option<String> {
+        self.0.as_str().map(|s| s.to_string())
+    }
+}
+
+#[derive(Clone)]
+pub struct SessionData {
+    pub value: HashMap<DataKey, DataValue>,
+    created_at: Instant,
+}
+
+impl Default for SessionData {
+    fn default() -> Self {
+        SessionData {
+            value: HashMap::new(),
+            created_at: Instant::now(),
+        }
+    }
+}
+
+pub struct InMemoryGatewaySession {
```

Comment:
I don't believe there is ever a reason to have an in-memory version of this. Let's build this directly on top of the key-value store (and add a ttl there)

Reply:
Removed but not committed yet

Reply:
We should have a in-memory session (sort of acting as a cache) on top of real kv store